walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,113 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() <= TimeRange.WHOLE_DAY.duration()){
+        ArrayList<TimeRange> mandatoryTimes = findAvailableMeetingTimes(events, request, true);
+        ArrayList<TimeRange> optionalTimes =  findAvailableMeetingTimes(events,request, false);
+        Collection<String> mandatoryAttendees = request.getAttendees();
+        Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+        if(optionalAttendees.size()== 0){
+            return mandatoryTimes;
+        }
+        if(mandatoryAttendees.size()== 0){
+            return optionalTimes;
+        }
+        ArrayList<TimeRange> commonTimes = findCommonTimes(mandatoryTimes,optionalTimes, request);
+        if(commonTimes.size() == 0){
+            return mandatoryTimes;
+        }
+        else{
+            return commonTimes;
+        }
+    
+    }
+    else {
+      return Collections.emptyList();
+    }
+  }
+  
+  //for isMandatoryMeeting, true means mandatory attendees and false means optional attendees
+   public ArrayList<TimeRange> findAvailableMeetingTimes(Collection<Event> events, MeetingRequest request, boolean findTimesforMandatoryAttendees){
+      ArrayList<TimeRange> meetingTimes = new ArrayList<TimeRange>(Arrays.asList(TimeRange.WHOLE_DAY));
+      Collection<String> meetingAttendees = findTimesforMandatoryAttendees ? request.getAttendees() : request.getOptionalAttendees();
+      for (Event event : events){
+        Set<String> eventAttendees = event.getAttendees();
+        if (Collections.disjoint(meetingAttendees, eventAttendees) ==  false){ 
```

Comment:
Explicit boolean comparison is usually not used, and IMO harder to read than simple negation. `if (!Collections.disjoint(...` reads as "if the collections are not disjoint", whereas `if (Collections.disjoint(...) == false` reads as "if the collections are disjoint is false" which is slightly harder to reason about. That's partly because `== true` or `== false` is not a thing people usually do in practice, so I'm not used to it, but partly because it's introducing extra boolean values (here an extra value of "false") and that means you then have to mentally reconcile two distinct values rather than just transforming a single boolean value to understand what's going on.

When the result of an operation is already a boolean, there's almost never a need to use an equality operation on it. The equality operator takes pairs of values and turns them into booleans, so if you already have a boolean, most of the time there's no need to do anything besides 'and', 'or', and 'not' to reach the intended condition. The one exception is `x != y`, but that's because `!=` for booleans is the same as 'xor', and you would never do that unless both sides are variables.

Consider for example that you could go one step further and write `if ((Collections.disjoint(meetingAttendees, eventAttendees) ==  false) == true) {`. This *also* produces the same result, but now there are three boolean values to think about when trying to read it and undersand what it does.

Reply:
fixed